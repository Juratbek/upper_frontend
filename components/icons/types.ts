export interface IIconProps {
  color?: string;
  variant?: 'outlined' | 'fulfilled';
  width?: number;
  height?: number;
  opacity?: number;
}
