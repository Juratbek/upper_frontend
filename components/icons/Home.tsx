import { IIconProps } from './types';

export const HomeIcon = ({ color = 'black', width = 20, height = 20 }: IIconProps) => (
  <svg width={width} height={height} viewBox='0 0 24 24' fill='none'>
    <path
      d='M9 17V21.9997H5C3.89543 21.9997 3 21.1043 3 19.9997V10.0332C3 9.73801 3.13046 9.45786 3.35643 9.26786L10.7129 3.08228C11.4569 2.45664 12.5431 2.45665 13.2871 3.08228L20.6436 9.26786C20.8695 9.45786 21 9.73801 21 10.0332V20C21 21.1046 20.1046 22 19 22H15V17C15 13 9 13 9 17Z'
      fill={color}
    />
  </svg>
);
