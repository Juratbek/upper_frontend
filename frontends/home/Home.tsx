import { ApiErrorBoundary } from 'components';
import { Spinner } from 'components/lib';
import {
  DefaultLabel,
  LABEL_ID_PARAM,
  LoadMoreButton,
  NoArticle,
  PublishedArticle,
  UserLabels,
} from 'components/molecules';
import { Advertisement } from 'components/organisms';
import { useDevice, useUrlParams } from 'hooks';
import { FC } from 'react';
import { usePublishedArticlesList } from 'store/clients/published-article';
import { addAmazonBucketUrl } from 'utils/published-article';

export const HomePage: FC = () => {
  const { getParam } = useUrlParams();
  const { isDesktop } = useDevice();
  const label = (getParam(LABEL_ID_PARAM) ?? DefaultLabel) as string;
  const {
    fetchNextPage,
    list: articles,
    hasNextPage,
    ...articlesRes
  } = usePublishedArticlesList(label);

  return (
    <>
      <UserLabels />
      {!isDesktop && <Advertisement />}
      <ApiErrorBoundary
        res={articlesRes}
        fallback={
          <div style={{ height: '15rem' }} className='content-center'>
            <Spinner />
          </div>
        }
      >
        {articles.length === 0 && !articlesRes.isLoading && <NoArticle label={label} />}
        {articles.map(addAmazonBucketUrl).map((article) => (
          <PublishedArticle article={article} key={article.id} />
        ))}
        {articles.length !== 0 && hasNextPage && (
          <LoadMoreButton
            onClick={() => fetchNextPage()}
            loading={articlesRes.isFetchingNextPage}
          />
        )}
      </ApiErrorBoundary>
    </>
  );
};
