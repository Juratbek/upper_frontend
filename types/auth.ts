import { Profile } from 'next-auth';
import { BuiltInProviderType } from 'next-auth/providers';
import { ClientSafeProvider, LiteralUnion } from 'next-auth/react';
import { AUTH_PROVIDER_TYPES } from 'variables';

type TGitHubUserType = 'User';

export type TAuthProviderProp = Profile & Record<string, unknown>;

export type TAuthProviderTypes =
  | typeof AUTH_PROVIDER_TYPES.google
  | typeof AUTH_PROVIDER_TYPES.facebook
  | typeof AUTH_PROVIDER_TYPES.github;

export type TAuthProviders = Record<LiteralUnion<BuiltInProviderType, string>, ClientSafeProvider>;

export interface IAuthProvider {
  [name: string]: (profile: TAuthProviderProp) => boolean;
}

export interface IGitHubUser {
  login: string;
  id: number;
  node_id: string;
  avatar_url: string;
  gravatar_id: string;
  url: string;
  html_url: string;
  followers_url: string;
  following_url: string;
  gists_url: string;
  starred_url: string;
  subscriptions_url: string;
  organizations_url: string;
  repos_url: string;
  events_url: string;
  received_events_url: string;
  type: TGitHubUserType;
  site_admin: boolean;
  name?: string;
  company?: string;
  blog: string;
  location?: string;
  email?: string;
  hireable?: boolean;
  bio?: string;
  twitter_username?: string;
  public_repos: number;
  public_gists: number;
  followers: number;
  following: number;
  created_at: string;
  updated_at: string;
}

export interface IGoogleUser {
  iss: string;
  azp: string;
  aud: string;
  sub: string;
  email: string;
  email_verified: boolean;
  at_hash: string;
  name: string;
  picture: string;
  given_name: string;
  family_name: string;
  locale: string;
  iat: number;
  exp: number;
}

export interface IFacebookUser {
  id: string;
  name: string;
  email: string;
  picture: {
    data: {
      height: number;
      is_silhouette: boolean;
      url: string;
      width: number;
    };
  };
}

export interface ITelegramUser {
  auth_date: number;
  first_name: string;
  last_name?: string;
  photo_url: string;
  hash: string;
  id: number;
  username: string;
}
