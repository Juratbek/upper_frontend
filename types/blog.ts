import { ILink } from './common';

export interface IBlogSmall {
  id: number;
  imgUrl: string;
  name: string;
}
export interface IBlogMedium extends IBlogSmall {
  bio?: string;
  followersCount?: number;
  articlesCount?: number;
}

export interface IBlog extends IBlogMedium {
  links: ILink[];
  createdDate: string;
  isFollowed?: boolean;
}
