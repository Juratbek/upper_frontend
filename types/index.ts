export * from './common';
export * from './article';
export * from './blog';
export * from './label';
export * from './notification';
export * from './auth';
