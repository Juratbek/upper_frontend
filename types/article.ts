import { OutputData } from '@editorjs/editorjs';

import { IBlog } from './blog';
import { ILabel } from './label';

export interface IArticleResult {
  id: number;
  title: string;
  content: string;
  imgUrl?: string;
  author?: IBlog;
  labels: ILabel[];
  publishedDate: Date;
  updatedDate?: Date;
  viewCount?: number;
  likes?: number;
}

export interface IArticle {
  id: number;
  title: string;
  content: OutputData;
  imgUrl?: string;
  author?: IBlog;
  labels: ILabel[];
  publishedDate: string;
  updatedDate?: string;
  viewCount?: number;
  likes?: number;
  prevArticleId?: number;
  nextArticleId?: number;
}
