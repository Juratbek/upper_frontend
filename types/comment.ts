import { IBlogSmall } from './blog';

export interface IComment {
  id: number;
  date: string;
  text: string;
  author: IBlogSmall;
}
