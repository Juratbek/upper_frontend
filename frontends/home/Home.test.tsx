import { render } from '@testing-library/react';
import { QueryWrapper, ReduxWrapper } from 'tests';
import { IPublishedArticleItem } from 'types';
import { vi } from 'vitest';

import { HomePage } from './Home';

const mockAuthor: IPublishedArticleItem['author'] = {
  id: 12,
  imgUrl: 'https://image.upper.uz/12',
  name: "Jur'atbek",
};

const mockArticles: IPublishedArticleItem[] = [
  {
    id: 1,
    imgUrl: 'https://image.upper.uz/1',
    title: 'Published article',
    author: mockAuthor,
  },
  {
    id: 2,
    imgUrl: 'https://image.upper.uz/2',
    title: 'Second published article',
    author: mockAuthor,
  },
  {
    id: 3,
    imgUrl: '',
    title: 'Third published article',
    author: mockAuthor,
  },
];

const mocks = vi.hoisted(() => ({
  usePublishedArticlesList: vi.fn().mockReturnValue({
    fetchNextPage: vi.fn(),
    list: [],
    hasNextPage: false,
    isLoading: true,
    isPending: true,
  }),
}));

vi.mock('store/clients/published-article', () => ({
  usePublishedArticlesList: mocks.usePublishedArticlesList,
}));

const Home = (
  <ReduxWrapper>
    <QueryWrapper>
      <HomePage />
    </QueryWrapper>
  </ReduxWrapper>
);

describe('Home', () => {
  it('snapshot', () => {
    render(Home);

    expect(document.body).toMatchSnapshot();
  });

  it('snapshot with articles returned by backend', () => {
    mocks.usePublishedArticlesList.mockReturnValue({
      list: mockArticles,
      fetchNextPage: vi.fn(),
      isLoading: false,
      isPending: false,
      isSuccess: true,
      hasNextPage: true,
    });

    render(Home);
    expect(document.body).toMatchSnapshot();
  });

  it('snapshot with empty articles returned by backend', () => {
    mocks.usePublishedArticlesList.mockReturnValue({
      list: [],
      fetchNextPage: vi.fn(),
      hasNextPage: false,
      isLoading: false,
      isPending: false,
      isSuccess: true,
    });

    render(Home);
    expect(document.body).toMatchSnapshot();
  });
});
