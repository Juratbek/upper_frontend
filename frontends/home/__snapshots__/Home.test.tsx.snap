// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home > snapshot 1`] = `
<body>
  <div>
    <div>
      <div
        class="root"
        style="padding-top: 2rem; margin-top: -2rem;"
      >
        <div
          class="container"
        >
          <div
            class="labels-container"
            id="labels"
          >
            <div
              class="lebels-list"
            >
              <button
                class="button primary"
                color="primary"
              >
                Eng so'nggi
              </button>
              <button
                class="button outlined"
                color="outlined"
              >
                Top
              </button>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="content-center"
          style="height: 15rem;"
        >
          <div
            class="spinner spinner--primary"
            role="spinner"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Home > snapshot with articles returned by backend 1`] = `
<body>
  <div>
    <div>
      <div
        class="root"
        style="padding-top: 2rem; margin-top: -2rem;"
      >
        <div
          class="container"
        >
          <div
            class="labels-container"
            id="labels"
          >
            <div
              class="lebels-list"
            >
              <button
                class="button primary"
                color="primary"
              >
                Eng so'nggi
              </button>
              <button
                class="button outlined"
                color="outlined"
              >
                Top
              </button>
            </div>
          </div>
        </div>
      </div>
      <div>
        <a
          class="article link"
          href="/web/articles/1"
        >
          <div
            class="article-img article-img article-img--full"
          >
            <img
              class="img"
              src="https://image.upper.uz/1"
            />
          </div>
          <h2
            class="article-title"
          >
            Published article
          </h2>
          <div
            class="article-footer"
          >
            <a
              class="link link"
              href="/web/blogs/12"
            >
              <div
                class="blog"
              >
                <div
                  class="avatar avatar--micro avatar"
                  style="border-color: rgba(245, 245, 245, 0.1);"
                >
                  <img
                    alt="UPPER"
                    data-nimg="fill"
                    decoding="async"
                    loading="lazy"
                    sizes="100vw"
                    src="http://localhost:3000/_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=3840&q=75"
                    srcset="/_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=3840&q=75 3840w"
                    style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: cover; color: transparent;"
                  />
                </div>
                <h5
                  class="name"
                >
                  Jur'atbek
                </h5>
              </div>
            </a>
          </div>
        </a>
        <a
          class="article link"
          href="/web/articles/2"
        >
          <div
            class="article-img article-img article-img--full"
          >
            <img
              class="img"
              src="https://image.upper.uz/2"
            />
          </div>
          <h2
            class="article-title"
          >
            Second published article
          </h2>
          <div
            class="article-footer"
          >
            <a
              class="link link"
              href="/web/blogs/12"
            >
              <div
                class="blog"
              >
                <div
                  class="avatar avatar--micro avatar"
                  style="border-color: rgba(245, 245, 245, 0.1);"
                >
                  <img
                    alt="UPPER"
                    data-nimg="fill"
                    decoding="async"
                    loading="lazy"
                    sizes="100vw"
                    src="http://localhost:3000/_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=3840&q=75"
                    srcset="/_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=3840&q=75 3840w"
                    style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: cover; color: transparent;"
                  />
                </div>
                <h5
                  class="name"
                >
                  Jur'atbek
                </h5>
              </div>
            </a>
          </div>
        </a>
        <a
          class="article link"
          href="/web/articles/3"
        >
          <h2
            class="article-title"
          >
            Third published article
          </h2>
          <div
            class="article-footer"
          >
            <a
              class="link link"
              href="/web/blogs/12"
            >
              <div
                class="blog"
              >
                <div
                  class="avatar avatar--micro avatar"
                  style="border-color: rgba(245, 245, 245, 0.1);"
                >
                  <img
                    alt="UPPER"
                    data-nimg="fill"
                    decoding="async"
                    loading="lazy"
                    sizes="100vw"
                    src="http://localhost:3000/_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=3840&q=75"
                    srcset="/_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fimage.upper.uz%2F12&w=3840&q=75 3840w"
                    style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: cover; color: transparent;"
                  />
                </div>
                <h5
                  class="name"
                >
                  Jur'atbek
                </h5>
              </div>
            </a>
          </div>
        </a>
        <button
          class="button"
          color="tertiary"
          style="opacity: 1; font-size: 20px; width: 100%;"
        >
          Yana yuklash
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`Home > snapshot with empty articles returned by backend 1`] = `
<body>
  <div>
    <div>
      <div
        class="root"
        style="padding-top: 2rem; margin-top: -2rem;"
      >
        <div
          class="container"
        >
          <div
            class="labels-container"
            id="labels"
          >
            <div
              class="lebels-list"
            >
              <button
                class="button primary"
                color="primary"
              >
                Eng so'nggi
              </button>
              <button
                class="button outlined"
                color="outlined"
              >
                Top
              </button>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div>
          <h3>
            Eng so'nggi
             uchun maqolalar topilmadi
          </h3>
          <p>
            Eng so'nggi
             mavzusida maqolalar yozing va bilim ulashing
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;
